### PR TITLE
Fix parallel collection inserts for PostgreSQL

### LINKED ISSUE #1488: Concurrent creation of database for parallel inserts doesn't work when database doesn't exist
### Versions

Nov 17, `main`

### What did you do?

Run dance tests and got an internal error

```
document_validation_test.go:112: 
        	Error Trace:	/home/user/dance/tests/diff/document_validation_test.go:112
        	Error:      	Received unexpected error:
        	            	(InternalError) [pool.go:213 pgdb.(*Pool).InTransaction] [error.go:200 common.CheckError] [insert.go:44 pgdb.InsertDocument] database/schema or collection/table already exist
        	Test:       	TestDocumentValidation/Update/DollarSign
        --- FAIL: TestDocumentValidation/Update/DollarSign (0.03s)
```

### What did you expect to see?

No errors

### What did you see instead?

DB already exists error.

### Reasoning

This is what `CreateDatabaseIfNotExists` promises (` If the schema already exists, no error is returned`):

```go
// CreateDatabaseIfNotExists creates a new FerretDB database (PostgreSQL schema).
// If the schema already exists, no error is returned, and transaction is not aborted.
func CreateDatabaseIfNotExists(ctx context.Context, tx pgx.Tx, db string) error {
```

This is what this function could return in fact:

```go
	case pgerrcode.DuplicateSchema:
		return ErrAlreadyExist
	case pgerrcode.UniqueViolation, pgerrcode.DuplicateObject:
		// https://www.postgresql.org/message-id/CA+TgmoZAdYVtwBfp1FL2sMZbiHCWT4UPrzRLNnX1Nb30Ku3-gg@mail.gmail.com
		// The same thing for schemas. Reproducible by integration tests.
		return ErrAlreadyExist
```

In addition to that, the statement that  `transaction is not aborted` is false. If `PgError` is caught, the transaction *will be* aborted (`current transaction is aborted, commands ignored until end of transaction block`).

### TODO

- [x] If possible, add a stress test to reproduce the problem on FerrretDB itself (e.g. parallel insert into the same database that doesn't exist)
- [x]  Fix the problem (e.g. fix godoc, check for `ErrAlreadyExists` in callers, use multiple transactions when needed - like [here](https://github.com/FerretDB/FerretDB/blob/af2554ad143a40afb110b4277995507af7873f73/internal/handlers/pg/msg_create.go#L75))
- [ ] Check dance for the current issue being mentioned and fix relevant TODOs

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #1488.

## Readiness checklist

* [x] I added tests for new functionality or bugfixes.
* [ ] I ran `task all`, and it passed.
* [ ] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

**Review by rumyantseva — CHANGES_REQUESTED** [review]
Tigris tests fail because Tigris needs some special creation-related handling. It might be relevant to this issue: #1341. Please check why Tigris fails, and if it's the same as #1341 and not easy to fix, add a todo with the issue number.

**Comment by rumyantseva** [comment]
@noisersup yeah, they started doing it, and we need to implement #1341 to support it correctly. So, for the current issue, I suggest to skip/todo the places about Tigris and concurrent collection creation and mentioning #1341.

**Review by w84thesun — APPROVED** [review]
LGTM

File: internal/handlers/pg/pgdb/collections.go [hunk 1]
```diff
@@ -144,8 +144,13 @@ func CreateCollection(ctx context.Context, tx pgx.Tx, db, collection string) err
 // If needed, it creates both database and collection.
 //
 // True is returned if collection was created.
-func CreateCollectionIfNotExist(ctx context.Context, tx pgx.Tx, db, collection string) (bool, error) {
-	exists, err := CollectionExists(ctx, tx, db, collection)
+func CreateCollectionIfNotExist(ctx context.Context, pgPool *Pool, db, collection string) (bool, error) {
```
Comment (chilagrow) [inline internal/handlers/pg/pgdb/collections.go:147]: This probably is outside the scope of this task, just noting that even if collection did exist when this func was called, nothing is preventing from collection to be deleted after calling this func.

**Review by chilagrow — CHANGES_REQUESTED** [review]

**Comment by noisersup** [comment]
As agreed on the call - error handling (rollbacking transaction) of the insert method will be improved on #940.

File: internal/handlers/pg/msg_insert.go [hunk 2]
```diff
@@ -63,22 +61,22 @@ func (h *Handler) MsgInsert(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
...
+		var doc any
+
+		doc, err = docs.Get(i)
+		if err != nil {
+			return nil, lazyerrors.Error(err)
 		}
-		return nil
-	})
+
+		err = h.insert(ctx, &sp, doc)
```
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/handlers/pg/msg_insert.go:73]: Nice catch!
https://github.com/FerretDB/FerretDB/pull/1513#issuecomment-1326212210

File: internal/handlers/pg/pgdb/collections.go (hunk 1 above)
Comment (noisersup) [reply internal/handlers/pg/pgdb/collections.go:147]: Yeah, things would be much easier if we could have collection creation and inserting in the same transaction

Comment (rumyantseva) [reply internal/handlers/pg/pgdb/collections.go:147]: As discussed in the call, an alternative approach could be using the same transaction and retrying it. The problem with it is with ordered inserts there are some errors that could be retried (from db and collection creation) and errors that couldn't (from inserts themselves). With such approach, we need to find a way to handle those retries correctly...

@noisersup let's wait for more review, but personally I'd vote for merging the current solutions with multiple transactions.

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by chilagrow — APPROVED** [review]

**Comment by AlekSi** [comment]
Please don't forget to enable auto-merge

**Pull request merged**